sdk/python/test_data/pipelines/pipeline_with_parallelfor_parallelism.py

```diff
@@ -24,6 +24,16 @@ def print_text(msg: str):
     print(msg)
 
 
+@component
+def print_int(x: int):
+    print(x)
+
+
+@component
+def list_dict_maker() -> List[Dict[str, int]]:
```

Comment:
This is a great update to this golden snapshot test and helps us assert the correctness of the full IR for a variety of different cases.

Do you think it also makes sense to discretely test each of the cases in a more typical unit test? This could go in [`compiler_test.py`](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/compiler/compiler_test.py).

I can think of the cases where the iterable is a:
1) Constant (no producer task annotation, since there is no producer task)
2) Upstream output and/or pipeline parameter with complete annotation `[List[Dict[str, int]]]`
3) Upstream output and/or pipeline parameter with partial annotation `[List[dict]]` or `[List[Dict]]`
4) Upstream output with a compiled + reloaded component (annotation is compressed to [STRUCT](https://github.com/kubeflow/pipelines/blob/0f3f68c05f620661abf4506504c80dc6646dc9a3/api/v2alpha1/pipeline_spec.proto#L434))

I think (3) and (4) are also omitted from this golden snapshot test.

While it's somewhat duplicative of this test, I think it might be helpful to explicitly ensure the fix is comprehensive by demonstrating that the pipeline compiles for each of these cases. I don't think we necessarily need to assert the content of the PipelineSpec IR for each case since we have this representative golden snapshot test.

Also, if your plan is to stage the fix for some of these case, that's totally fine.